pkg/streamingpromql/optimize/inspect.go

```diff
@@ -17,43 +18,50 @@ type InspectResult struct {
 	// IsRewrittenByMiddleware indicates if the query has been rewritten by query sharding or by
 	// subquery spin-off middlewares.
 	IsRewrittenByMiddleware bool
+	// CreatedLabels is a set of labels that are the result of a function that creates labels.
+	CreatedLabels map[string]struct{}
```

Comment:
I think we should check for labels from `label_replace` and `label_join` separately, once per binary operation, rather than for the overall expression.

Imagine the expression was something like `(foo * on (env, region) bar) * on (env, region) label_replace(baz, "region", "$1", "cluster", ".*")`. As it stands, we won't apply narrowing of `region` to either binary operation, even though it'd be safe to do so for the `foo` / `bar` binary operation.